block-layer/codecs/dag-jose.md

```diff
@@ -0,0 +1,126 @@
+# Specification: DAG-JOSE
+
+**Status: Descriptive - Draft**
+
+JOSE is a stanard for signing and encrypting JSON objects. The various specifications for JOSE can be found in the [IETF datatracker](https://datatracker.ietf.org/wg/jose/documents/). 
+
+DAG JOSE supports the full [IPLD Data Model](../data-model-layer/data-model.md) (within the payload).
+
+## Format
+
+The are two main ways to represent a JOSE object. As a JWS ([json web signature](https://datatracker.ietf.org/doc/rfc7515/?include_text=1)) and JWE ([json web encryption](https://datatracker.ietf.org/doc/rfc7516/?include_text=1)). These two formats  acts as the primitives in JOSE and can be used to create JWT and JWM objects etc. This specification describes how to encode JWS and JWE as an IPLD format.
+
+### Representation
+
+The layout of a decoded JOSE object is described by the IPLD schema defined below. We will refer to this layout as the `Decoded Representation`. 
+
+```ipldsch
+type Signature struct {
+  header optional {String:Any}
+  protected optional {String:Any}
+  signature Bytes
+}
+
+type JWS struct {
+  payload Any
+  signatures [Signature]
+}
+
+type Recipient struct {
+  encrypted_key optional Bytes
+  header optional {String:Any}
+}
+
+type JWE struct {
+  aad optional Bytes
+  ciphertext Bytes
+  iv optional Bytes
+  protected optional {String:Any}
+  recipients [Recipient]
+  tag optional Bytes
+  unprotected optional {String:Any}
+}
+
+type JOSE union {
+  | JWS jws
+  | JWE jwe
+} representation kinded
+```
+
+### Serialization
+
+Both JWS and JWE supports three different serialization formats: `Compact Serialization`, `Flattened JSON Serialization`, and `General JSON Serialization`. The first two are more concise, but they only allow for one recipient. Therefore DAG JOSE always uses the `General Serialization` which ensures maximum compatibility with minimum ambiguity. 
+
+The implementation of the serialization function should accept all JOSE formats including the `Decoded Representation` and convert them if necessary. 
+
+#### General JSON Serialization
+
+Below the  `General JSON Serialization` can be observed. Note that all data represented as `String` here is data that has been encoded using `base64url`. Converting `Compact Serialization` and `Flattened JSON Serialization` to the general serialization is trivial.
+
+```ipldsch
+type GeneralSignature struct {
+  header optional {String:Any}
+  protected optional String
+  signature String
+}
+
+type GeneralJWS struct {
+  payload String
+  signatures [GeneralSignature]
+}
+
+type GeneralRecipient struct {
+  encrypted_key optional String
+  header optional {String:Any}
+}
+
+type GeneralJWE struct {
+  aad optional String
+  ciphertext String
+  iv optional String
+  protected optional String
+  recipients [GeneralRecipient]
+  tag optional String
+  unprotected optional {String:Any}
+}
+
+type GeneralJOSE union {
```

Comment:
we have a discussion going on elsewhere, I think in the recently merged FlexibleByteLayout PR, that we should make sure that any entry points are listed at the top of a schema doc for clarity, something like "reverse C function definition order", i.e. list them as a top-down dependency tree where your root is where you _generally_ expect users to enter. In this case I think `GeneralJOSE` is your expected root type that everything falls under, so you probably need to invert both of your schema definitions. Of course accompanying this with some descriptive text is a nice touch too but probably optional.